_posts/2024-06-14-remote-code-execution-engine.md

```diff
@@ -0,0 +1,1426 @@
+
+        Behaviors.stopped
+```
+
+The details:
+- `Messages`: The `In` enum defines messages the actor can handle, including `PrepareFile`, `FilePrepared`, and `FilePreparationFailed`.
+- `Behavior Setup`: The `apply` method sets up the actor's behavior
+- On receiving `PrepareFile`, it logs the event, attempts to write the code to a file asynchronously, and uses `pipeToSelf` pattern to handle the result.
+- On `FilePrepared`, it spawns a `CodeExecutor` actor to execute the code, watches it for termination, and sends an `Execute` message to the executor.
+- On `FilePreparationFailed`, it logs the failure and sends an `ExecutionFailed` message back to the requester.
+ 
+And we arrived at the point where we can talk about the most important actor - `CodeExecutor`.
+
+## 5.3 CodeExecutor
+
+`CodeExecutor` actor will be a child of the local `FileHandler` actor on the same `worker` node.
+
+The main responsibility for `CodeExecutor` is to execute code and return the output.
+
+`CodeExecutor` actor is complex in nature due to the following reasons:
+- code must be executed in a short-lived docker container
+- code can cause memory issues
+- code can cause cpu issues
+- code can result in an accidental infinite loop
+- code can be accumulating too much output
+- and so on...
+
+To tackle all this, we need to write a special logic and model the aforementioned concepts in a sensible way.
+
+After struggling with those issues for a week or so, I managed to refine `CodeExecutor` iteratively, ending up with the following:
+
+```scala
+package workers.children
+
+import org.apache.pekko.actor.typed.ActorRef
+import org.apache.pekko.actor.typed.scaladsl.Behaviors
+import org.apache.pekko.stream.IOResult
+import org.apache.pekko.stream.scaladsl.{Sink, Source, StreamConverters}
+import org.apache.pekko.util.ByteString
+import workers.Worker
+
+import java.io.{File, InputStream}
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.*
+import scala.util.control.NoStackTrace
+
+object CodeExecutor:
+
+  private val KiloByte = 1024 // 1024 bytes
+  private val MegaByte = KiloByte * KiloByte // 1,048,576 bytes
+  private val TwoMegabytes = 2 * MegaByte // 2,097,152 bytes
+
+  private val AdjustedMaxSizeInBytes =
+    (TwoMegabytes * 20) / 100 // 419,430 bytes, which is approx 409,6 KB or 0.4 MB
+
+  // max size of output when the code is run, if it exceeds the limit then we let the user know to reduce logs or printing
+  private val MaxOutputSize = AdjustedMaxSizeInBytes
+
+  enum In:
+    case Execute(compiler: String, file: File, dockerImage: String, replyTo: ActorRef[Worker.In])
+    case Executed(output: String, exitCode: Int, replyTo: ActorRef[Worker.In])
+    case ExecutionFailed(why: String, replyTo: ActorRef[Worker.In])
+    case ExecutionSucceeded(output: String, replyTo: ActorRef[Worker.In])
+
+  private case object TooLargeOutput extends Throwable with NoStackTrace {
```

Comment:
This is one more reason to stick to the brace-based syntax. By reflex, we tend to mix the styles, which is worse than using either style.